File: App/Profiles/ProfileViewModel.m
//  ProfileViewModel.m
//
//  Copyright 2013 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

#import "ProfileViewModel.h"
#import "AwfulSettings.h"
#import "Awful-Swift.h"

@interface ProfileViewModel ()

@property (nonatomic) Profile *profile;

@end

@implementation ProfileViewModel

- (instancetype)initWithProfile:(Profile *)profile
{
    if ((self = [super init])) {
        _profile = profile;
    }
    return self;
}

- (instancetype)init
{
    NSAssert(nil, @"Use -initWithProfile: instead");
    return [self initWithProfile:nil];
}

- (NSString *)stylesheet
{
    NSURL *URL = [[NSBundle mainBundle] URLForResource:@"profile" withExtension:@"css"];
    NSError *error;
    NSString *stylesheet = [NSString stringWithContentsOfURL:URL encoding:NSUTF8StringEncoding error:&error];
    if (!stylesheet) {
        NSLog(@"%s error loading stylesheet from %@: %@", __PRETTY_FUNCTION__, URL, error);
    }
    return stylesheet;
}

- (NSString *)userInterfaceIdiom
{
    return UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad ? @"ipad" : @"iphone";
}

- (BOOL)dark
{
    return [AwfulSettings sharedSettings].darkTheme;
}

- (NSDateFormatter *)regDateFormat
{
    return [NSDateFormatter regDateFormatter];
}

- (NSDateFormatter *)lastPostDateFormat
{
	return [NSDateFormatter postDateFormatter];
}

- (BOOL)anyContactInfo
{
    if (self.privateMessagesWork) return YES;
    NSDictionary *contactInfo = [self.profile dictionaryWithValuesForKeys:@[ @"aimName", @"icqName", @"yahooName", @"homepageURL" ]];
    for (id value in contactInfo.allValues) {
        if ([value respondsToSelector:@selector(length)] && [value length] > 0) return YES;
        if ([value respondsToSelector:@selector(absoluteString)] && [value absoluteString].length > 0) return YES;
    }
    return NO;
}

- (BOOL)privateMessagesWork
{
    return self.profile.user.canReceivePrivateMessages && [AwfulSettings sharedSettings].canSendPrivateMessages;
}

- (NSString *)customTitleHTML
{
    NSString *HTML = self.profile.user.customTitleHTML;
    return [HTML isEqualToString:@"<br/>"] ? nil : HTML;
}

- (NSString *)gender
{
    return self.profile.gender ?: @"porpoise";
}

- (NSURL *)avatarURL
{
    return self.profile.user.avatarURL;
}

- (NSDate *)regdate
{
    return self.profile.user.regdate;
}

- (NSString *)username
{
    return self.profile.user.username;
}

- (id)valueForUndefinedKey:(NSString *)key
{
    return [self.profile valueForKey:key];
}

@dynamic aboutMe;
@dynamic aimName;
@dynamic avatarURL;
@dynamic homepageURL;
@dynamic icqName;
@dynamic interests;
@dynamic lastPost;
@dynamic location;
@dynamic occupation;
@dynamic postCount;
@dynamic postRate;
@dynamic profilePictureURL;
@dynamic regdate;
@dynamic username;
@dynamic yahooName;

@end

File: App/Profiles/ProfileViewModel.swift
//  ProfileViewModel.swift
//
//  Copyright 2016 Awful Contributors. CC BY-NC-SA 3.0 US https://github.com/Awful/Awful.app

import AwfulCore
import Mustache

struct ProfileViewModel: MustacheBoxable {
    private let dict: [String: Any]
    
    init(_ profile: Profile) {
        var privateMessagesWork: Bool {
            guard profile.user.canReceivePrivateMessages else { return false }
            return AwfulSettings.shared().canSendPrivateMessages
        }
        var anyContactInfo: Bool {
            if privateMessagesWork { return true }
            if let aim = profile.aimName, !aim.isEmpty { return true }
            if let icq = profile.icqName, !icq.isEmpty { return true }
            if let yahoo = profile.yahooName, !yahoo.isEmpty { return true }
            if profile.homepageURL != nil { return true }
            return false
        }
        var customTitleHTML: String? {
            guard let html = profile.user.customTitleHTML, html != "<br/>" else { return nil }
            return html
        }

        dict = [
            "aboutMe": profile.aboutMe as Any,
            "aimName": profile.aimName as Any,
            "anyContactInfo": anyContactInfo,
            "avatarURL": profile.user.avatarURL as Any,
            "customTitleHTML": customTitleHTML as Any,
            "dark": AwfulSettings.shared().darkTheme,
            "gender": profile.gender ?? "porpoise",
            "homepageURL": profile.homepageURL as Any,
            "icqName": profile.icqName as Any,
            "interests": profile.interests as Any,
            "lastPost": profile.lastPostDate as Any,
            "location": profile.location as Any,
            "occupation": profile.occupation as Any,
            "postCount": Int(profile.postCount),
            "postRate": profile.postRate as Any,
            "privateMessagesWork": privateMessagesWork,
            "profilePictureURL": profile.profilePictureURL as Any,
            "regdate": profile.user.regdate as Any,
            "stylesheet": loadStylesheet(),
            "username": profile.user.username as Any,
            "yahooName": profile.yahooName as Any]
    }

    var mustacheBox: MustacheBox {
        return Box(dict)
    }
}

private class BundleLocator {}

private func loadStylesheet() -> String {
    guard let url = Bundle(for: BundleLocator.self)
        .url(forResource: "profile.css", withExtension: nil)
        else { fatalError("missing profile.css") }

    do {
        return try String(contentsOf: url, encoding: String.Encoding.utf8)
    } catch {
        fatalError("couldn't load \(url): \(error)")
    }
}
